src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -73,167 +74,96 @@ public override void Accept(OperationVisitor visitor)
+                                                : null)
+                                            : (argumentsToParametersOpt[a] > 0
+                                                ? parameters[argumentsToParametersOpt[a]]
+                                                : null);
+
+                            argumentsWithErrors.Add(new Argument(ArgumentKind.Explicit, parameter, boundArguments[a]));
                         }
 
-                        if (argumentIndex >= parameters.Length - 1 &&
-                            parameters.Length > 0 &&
-                            parameters[parameters.Length - 1].IsParams &&
-                            // An argument that is an array of the appropriate type is not a params argument.
-                            (boundArguments.Length > argumentIndex + 1 ||
-                             ((object)argument.Type != null && // If argument type is null, we are in an error scenario and cannot tell if it is a param array, or not. 
-                              (argument.Type.TypeKind != TypeKind.Array ||
-                              !argument.Type.Equals(parameters[parameters.Length - 1].Type, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)))))
-                        {
-                            return new Argument(ArgumentKind.ParamArray, parameters[parameters.Length - 1], CreateParamArray(parameters[parameters.Length - 1], boundArguments, argumentIndex, invocationSyntax));
-                        }
-                        else
-                        {
-                            return new SimpleArgument(parameterOpt, argument);
-                        }
+                        return argumentsWithErrors.ToImmutableAndFree();
+                    } 
+
+                    var argumentKindBuilder = ArrayBuilder<ArgumentKind>.GetInstance(parameters.Length);
+                    var parameterSymbolBuilder = ArrayBuilder<ParameterSymbol>.GetInstance(parameters.Length);
+
+                    var derivedArguments = LocalRewriter.MakeArgumentsInEvaluationOrder(
+                        binder: binder,
+                        syntax: invocationSyntax,
+                        arguments: boundArguments,
+                        methodOrIndexer: methodOrIndexer,
+                        optionalParametersMethod: optionalParametersMethod,
+                        expanded: expanded,
+                        argsToParamsOpt: argumentsToParametersOpt,
+                        invokedAsExtensionMethod: invokedAsExtensionMethod,
+                        argumentKindBuilder: argumentKindBuilder,
+                        parameterSymbolBuilder: parameterSymbolBuilder);
+                    
+                    Debug.Assert(derivedArguments.Length == parameters.Length 
+                              && derivedArguments.Length == argumentKindBuilder.Count
+                              && derivedArguments.Length == parameterSymbolBuilder.Count);
```

Comment:
Consider splitting into three separate asserts so it's clear which expression failed when there is a failure.